python/tests/test_vecadd.py

```diff
@@ -1,79 +1,222 @@
     def kernel(x_ptr,
                y_ptr,
                z_ptr,
-               BLOCK_SIZE: tl.constexpr):
+               block_size,
+               iter_size: tl.constexpr):
         pid = tl.program_id(axis=0)
-        offset = pid * BLOCK_SIZE + tl.arange(0, BLOCK_SIZE)
-        x_ptrs = x_ptr + offset
-        y_ptrs = y_ptr + offset
-        x = tl.load(x_ptrs)
-        y = tl.load(y_ptrs)
-        z = x + y
-        z_ptrs = z_ptr + offset
-        tl.store(z_ptrs, z)
+        for i in range(0, block_size, iter_size):
+            offset = pid * block_size + tl.arange(0, iter_size)
+            x_ptrs = x_ptr + offset
+            y_ptrs = y_ptr + offset
 
-    x = torch.randn((BLOCK_SIZE,), device='cuda', dtype=torch.float32)
-    y = torch.randn((BLOCK_SIZE,), device='cuda', dtype=torch.float32)
-    z = torch.empty((BLOCK_SIZE,), device=x.device, dtype=x.dtype)
+            x = tl.load(x_ptrs)
+            y = tl.load(y_ptrs)
+            z = x + y
+            z_ptrs = z_ptr + offset
+            tl.store(z_ptrs, z)
+
+            x_ptr += iter_size
+            y_ptr += iter_size
+            z_ptr += iter_size
 
-    grid = lambda EA: (x.shape.numel() // BLOCK_SIZE,)
-    kernel[grid](x_ptr=x, y_ptr=y, z_ptr=z, BLOCK_SIZE=BLOCK_SIZE, num_warps=NUM_WARPS)
+    x = torch.randn((block_size,), device='cuda', dtype=torch.float32)
+    y = torch.randn((block_size,), device='cuda', dtype=torch.float32)
+    z = torch.empty((block_size,), device=x.device, dtype=x.dtype)
+
+    grid = lambda EA: (x.shape.numel() // (block_size),)
+    kernel[grid](x_ptr=x, y_ptr=y, z_ptr=z,
+                 block_size=x.shape[0], iter_size=iter_size, num_warps=num_warps)
 
     golden_z = x + y
-    assert_allclose(z, golden_z, rtol=1e-7, atol=1e-7)
+    assert_close(z, golden_z, rtol=1e-7, atol=1e-7)
 
 
-@pytest.mark.parametrize('NUM_WARPS, BLOCK_SIZE, ITER_SIZE', [
-    [4, 256, 1],
-    [4, 1024, 256],
+@pytest.mark.parametrize('shape, num_warps, block_size, iter_size', [
+    [(127, 3), 2, 128, 1],
+    [(127, 3), 2, 128, 32],
 ])
-def test_vecadd_scf_no_mask(NUM_WARPS, BLOCK_SIZE, ITER_SIZE):
-
+def test_vecadd_scf_mask(shape, num_warps, block_size, iter_size):
     @triton.jit
     def kernel(x_ptr,
                y_ptr,
                z_ptr,
-               BLOCK_SIZE,
-               ITER_SIZE: tl.constexpr):
+               num_elements,
+               block_size: tl.constexpr,
+               iter_size: tl.constexpr
+               ):
+        '''
+        @block_size: size of a block
+        @iter_size: size of the iteration, a block has multiple iterations
+        @num_elements: number of elements
+        '''
         pid = tl.program_id(axis=0)
-        for i in range(0, BLOCK_SIZE, ITER_SIZE):
-            offset = pid * BLOCK_SIZE + tl.arange(0, ITER_SIZE)
+        for i in range(math.ceil(block_size / iter_size)):
+            # TODO: a bug here, if put the offset outside the forloop, there will be a GPU mis-aligned error.
```

Comment:
hmmm..this sounds like a critical bug that worth our attention. I was not aware of this when i put this line in the for loop. Did you see anything suspicious in the IR when the bug is reproduced?

Reply:
Yeah, I can reproduce it in this case but I haven't dug in this issue.